### PR TITLE
Arikt/usage counting

### PR DESCRIPTION
**Pull request by arikt-ms**
# Adding usage counting for signals and client connectivity.

For more information about how to contribute to this repo, visit this [page](https://github.com/microsoft/FluidFramework/blob/main/CONTRIBUTING.md)

## Description

As a paid service FRS (Azure Fluid Relay) needs to have a way to meter the usage of certain resources defined per business plan. Those resources include the number of incoming and outgoing signals and the client connection time in minutes, among others. The metering of these resources requires counting the corresponding usage. Given that both resources are associated with socket activity, the Alfred lambda naturally would be the most appropriate place to add the usage counting code. On the other hand, given that socket signal handler is on the critical path, the usage counting should be done in a way to minimize the perf impact on the critical path. For that matter, it has been decided to re-use the existing throttler infra as much as possible to avoid introducing any new functionality on the critical path. The following changes have been made to enable the usage counting for signals and client connection time:

1. Existing throttler code has been extended to allow the passing and further storage of counted usage data.
2. Throttling has been added/enabled for signals.
3. With updated throttler code, the counted signal usage will be stored to Redis with every invocation of throttler algorithm. The storage of counted usage data is combined with the storage of throttling data to avoid introducing an extra network call.
4. The client connection time is calculated and stored to Redis upon client disconnection.
5. The usage data stored in Redis is expected to be picked up by a different component (not on critical path) for further processing per billing requirements.
6. Config switches for enabling/disabling the usage counting for both signals and client connection time (disabled by default).

## PR Checklist

-   [ ] I have updated the documentation accordingly.
-   [x] I have added tests to cover my changes.
-   [x] All new and existing tests passed.
-   [x] My code follows the code style of this project.
-   [x] I ran the lint checks which produced no new errors nor warnings for my changes.
-   [x] I have checked to ensure there aren't other open Pull Requests for the same update/change.

## Does this introduce a breaking change?

-   [ ] Yes
-   [x] No

## Testing

Changes have been validated both with a local Docker setup and deployed to a k8s cluster.

### CODE REVIEW

File: server/routerlicious/packages/lambdas/src/alfred/index.ts [hunk 1]
```diff
@@ -268,9 +310,13 @@ export function configureWebSocketServices(
             messageClient.user = claims.user;
             messageClient.scopes = claims.scopes;
 
-            // Do not give SummaryWrite scope to clients that are not summarizers
+             // 1. Do not give SummaryWrite scope to clients that are not summarizers.
```
Comment (znewton) [inline server/routerlicious/packages/lambdas/src/alfred/index.ts:313]: nit: there's an extra space in front of this comment

File: server/routerlicious/packages/lambdas/src/alfred/index.ts [hunk 2]
```diff
@@ -545,6 +613,18 @@ export function configureWebSocketServices(
                 );
                 // eslint-disable-next-line @typescript-eslint/no-floating-promises
                 connection.disconnect();
+                if (isClientConnectivityCountingEnabled && throttleAndUsageStorageManager) {
+                    const connectionTimestamp = connectionTimeMap.get(clientId);
+                    if (connectionTimestamp) {
+                        await storeClientConnectivityTime(
```
Comment (znewton) [inline server/routerlicious/packages/lambdas/src/alfred/index.ts:619]: This `await` in the for loop can/will be problematic. We would have to wait for the storage to occur between notifying the next client of a disconnect. I'd suggest pushing these to a promise array and awaiting outside the loop (see the removeP implementation a few lines down)

Comment (arikt-ms) [reply server/routerlicious/packages/lambdas/src/alfred/index.ts]: good point.

File: server/routerlicious/packages/lambdas/src/alfred/index.ts [hunk 3]
```diff
@@ -131,20 +133,54 @@ function parseRelayUserAgent(relayUserAgent: string | undefined): Record<string,
...
+ */
+ async function storeClientConnectivityTime(
+    clientId: string,
+    documentId: string,
+    tenantId: string,
+    connectionTimestamp: number,
+    throttleAndUsageStorageManager: core.IThrottleAndUsageStorageManager) {
+    try {
+        const now = Date.now();
+        const connectionTimeInMinutes = (now - connectionTimestamp) / 60000;
```
[the remark this answers is not in the document]
Comment (arikt-ms) [reply server/routerlicious/packages/lambdas/src/alfred/index.ts:147]: The measurement unit is defined by business plan.

**Review by sharmakhushboo — APPROVED** [review]

**Pull request merged**